src/main/java/org/jnode/fs/ntfs/datarun/CompressedDataRun.java

```diff
@@ -220,43 +248,96 @@ private static int uncompressBlock(final OffsetByteArray compressed,
             return 0;
         }
 
+        // Bit 15 indicates whether the chunk contains compressed data. If this bit is zero, the chunk header is followed by uncompressed literal data.
+        // If this bit is set, the next byte of the chunk is the beginning of a Flag_group nonterminal that describes some compressed data.
         if ((rawLen & 0x8000) == 0) {
             // Uncompressed chunks store length as 0xFFF always.
             if ((len + 1) != BLOCK_SIZE) {
                 log.debug("ntfs_uncompblock: len: " + len + " instead of 0xfff");
             }
 
-            // Copies the entire compression block as-is, need to skip the compression flag,
-            // no idea why they even stored it given that it isn't used.
-            // Darwin's version I was referring to doesn't skip this, which seems be a bug.
-            uncompressed.copyFrom(compressed, cpos, 0, len + 1);
-            uncompressed.zero(len + 1, BLOCK_SIZE - 1 - len);
-            cpos++;
+            for (int k = 0; k < len + 1; k++) {
+                if (uncompressed.offset + pos >= uncompressed.array.length ||
+                        compressed.offset + cpos >= compressed.array.length) {
+                    break;
+                }
+                uncompressed.put(pos++, compressed.get(cpos++));
+            }
+
+            // no need to put zero in the rest in uncompressed array.
             return len + 3;
         }
 
-        while (cpos < len + 3 && pos < BLOCK_SIZE) {
+        int rightmostInUncompressed = Math.min(uncompressed.array.length, BLOCK_SIZE);
+
+        // Now this chunk contains compressed data, decompress it.
+        while (cpos < len + 3 && pos < rightmostInUncompressed) {
+            // https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-xca/1fd21d29-f42f-4fc4-b677-9de7cc386be8
+            // If a chunk is compressed, its chunk header is immediately followed by the first byte of a Flag_group nonterminal.
```

Comment:
All the comments are copied from Microsoft official website. I feel it is a good idea to make the code crystal clear, though the comments make the code messy.